### DISCUSSION TITLE
Changing the Learning Assistant to PyTorch Lightning

### DISCUSSION THREAD

**Discussion by Barchid** [lava-nc/lava-dl#58] [Ideas]
Hello,

I am currently getting my hands dirty with the code (notably the Sigma-Delta Neural Net tutorial). I was wondering if there will be an advantage to replace the Learning Assistant (`slayer.utils.Assistant`) by a`pytorch-lightning` module. In this way, we can benefit from lots of built-in functionalities:
- Tensorboard Logger (instead of the `slayer.utils.LearningStats`)
- LR finder
- Easy Distribution on multiple devices (DDP strategies, ...) https://pytorch-lightning.readthedocs.io/en/latest/guides/speed.html#training-on-accelerators
- Mixed-precision Training https://pytorch-lightning.readthedocs.io/en/latest/guides/speed.html#mixed-precision-16-bit-training
- Hyper-params saving
- Auto Checkpoints 
- Easier to integrate other algorithms (e.g. Self-Supervised Learning with Lightning-Bolts)
- ......

### Example
I reimplemented the PilotNet-SDNN tutorial with `pytorch-lightning` if you want to see an example : https://github.com/Barchid/lava-dl-lightning . Some functionalities of the Assistant are still in progress (for example, the `count_log` argument).

In the project, the Lightning Module that can replace the Assistant is in `lit_lavadl.py`. Basically, it takes the parameters as the assistant and performs the same operations (i.e. integrates a `classifier`, a sparsity loss `lam` parameter, ..). In the script `main.py` you can see how you can use the module. There is a lot of code that is handled by 

What's cool with `lightning` is that you already have lots of program arguments to work with built-in properties. Type `python main.py --help` and you will have a quick look of the possibilities.

### An additional improvement on Sparsity Loss and Count_Log
In the SDNN tutorial (and Assistant), I saw that the sparsity loss and the event count metric are integrated directly in the network (i.e. you have to explicitly return these values in addition to the output). I think it would be easier if we don't integrate these kind of supervision/metric inside the network implementation. I brought a solution to avoid adding additional returned value by using pytorch hooks (see `sparsity_loss.py` in the project).

### What do you think?
If you think these changes can be beneficial for the library, I would be more than happy to contribute and integrate my changes. Let me know if I should work on that!

Cheers!
BARCHID Sami.

**Comment by bamsumit** [comment]
Hi @Barchid,

I like the `sparsity_loss` implementation. Event counter module using torch hooks also sounds great. It would be great if you can contribute them to __lava-dl__.

I was excited to see your port of lava-dl lightning. I just found it a couple of days ago. In principle, all lava-dl slayer and bootstrap modules are torch modules. So they should technically be usable with Pytorch lightning. `slayer.utils.Assistant` is just a lightweight alternative with a simple set of functionality.

What would be useful though is if you can contribute a notebook tutorial showing how to incorporate PyTorch lightning during the training. Ideally a new dataset would be great to add to the variety of examples. It's also fine if you do a parallel notebook to an existing one that shows PyTorch Lightning in use.

Regards,
Sumit

**Comment by Barchid** [comment]
Hi @bamsumit ,

Yes, I would be happy to bring some contributions :smile:  . I will definitely add a PR next week (when I have some time) about the `sparsity_loss` and `event_counter` modules.

I will add some lightning tutorials in another PR. I can definitely integrate a tutorial on gesture recognition with DVSGesture (using another popular neuromorphic library like [tonic](https://github.com/neuromorphs/tonic)). Don't hesitate if you have any idea about a tutorial you would like to see on the repository.

Kind regards,
BARCHID Sami.

**Comment by bamsumit** [comment]
Sounds good. DVS Gesture sounds great. Let me know if you need any help on the network architectures, augmentation etc.

**Comment by Barchid** [comment]
Thanks for your help! If you have any suggestion about a network architecture that works well with DVS Gesture, I take it.

Since I've seen the support of graded spike. That would be a great idea to show another different use case of `lava-dl` (and I'm quite interested for my own works). I would use them on DVS Gesture with summed event frames. However, I've never worked with graded spikes. Do you have any tips about good hyper-param choices, etc ?

**Comment by bamsumit** [comment]
Here is the architecture that I typically use `32x32x2-16c5s2-32c3s2-512d-11avg`.

* I also find that adding a sum pooling layer during training instead of downsampling also works better (`128x128x2-4a` before the above network config). But from hardware perspective, it's efficient to downsample the event to 32x32x2 and use it.
* Input augmentation is key. Typically I do a random shift of ±8 pixels and rotation of ±10 degrees.
* Training samples are random 300ms clips from the training set. There is an argument when converting event to tensor to enable random shift (see [here](https://lava-nc.org/lava-lib-dl/slayer/io.html#lava.lib.dl.slayer.io.Event.fill_tensor())).
* Testing samples are the first 1500ms of the testing set.

**Reply by Barchid** [reply]
Hello,
Just to give a few updates since it has been quite a long time:

The whole process works smoothly and the tutorial is written. However, I got poor results with defaut hyper-parameters I chose for the moment so I am currently tuning the hyperparameters. For short, my loss function is stuck to a certain value after one or two iterations.. I am surely doing something bad with my architecture :

```python
(....)
neuron_params = {
            'threshold': threshold,  # 1.25,
            'current_decay': current_decay,  # 0.25,
            'voltage_decay': voltage_decay,  # 0.03,
            'tau_grad': 1.,
            'scale_grad': 1.,
            'requires_grad': False,  # True,
            'graded_spike': False
        }
        neuron_params_drop = {
            **neuron_params,
            'dropout': slayer.neuron.Dropout(p=0.05),
        }
        neuron_params_drop = {**neuron_params}

        self.blocks = torch.nn.ModuleList([
            # reduce to 32x32x2 inputs
            slayer.block.cuba.Pool(neuron_params, kernel_size=4),
            slayer.block.cuba.Conv(
                neuron_params,
                2, 16, 5, padding=2, stride=2, weight_norm=True
            ),
            slayer.block.cuba.Conv(
                neuron_params,
                16, 32, 3, padding=1, stride=1, weight_norm=True
            ),
            slayer.block.cuba.Conv(
                neuron_params,
                32, 64, 3, padding=1, stride=2, weight_norm=True
            ),
            slayer.block.cuba.Conv(
                neuron_params,
                64, 128, 3, padding=1, stride=1, weight_norm=True
            ),
            slayer.block.cuba.Flatten(),
            slayer.block.cuba.Dense(
                neuron_params_drop,
                8 * 8 * 128, 512, weight_norm=True, weight_scale=weight_scale_dense
            ),
            slayer.block.cuba.Dense(
                neuron_params, 512, 11, weight_scale=weight_scale_dense,
                weight_norm=True
            ),
        ])
```

If someone has an idea, I would be more than happy to know it.

Cheers.

**Reply by bamsumit** [reply]
Hi @Barchid ,
What loss are you using? I would look at gradient flow of the network and the weight scale of the layers first.

**Reply by Barchid** [reply]
Thanks for helping me!

So, I applied some changes and I found something that characterizes the problem but I don't know a proper solution to this.

**What I use:**
- Loss: SpikeRate with true_rate=0.2 and false_rate=0.03 (I also tried true_rate=1. and false_rate=0.)
- Architecture:
```python
neuron_params = {
            'threshold'     : 1.25,
            'current_decay' : 0.25,
            'voltage_decay' : 0.03,
            'tau_grad'      : 1,
            'scale_grad'    : 1,
            'requires_grad' : False,
        }
        neuron_dense_params = {  # dense layers have additional dropout units enabled
            **neuron_params,                        # copy all sdnn_cnn_params
            'dropout': slayer.neuron.Dropout(p=0.2),  # neuron dropout
        }

        self.blocks = torch.nn.ModuleList([  # sequential network blocks
            slayer.block.cuba.Pool( # (2, 32, 32)
                neuron_params, 4
            ),
            slayer.block.cuba.Conv( # (16, 16, 16)
                neuron_params, 2, 16, 5, padding=2, stride=2, weight_scale=10, weight_norm=True
            ),
            slayer.block.cuba.Conv( # (32, 8, 8)
                neuron_params, 16, 32, 3, padding=1, stride=2, weight_scale=50, weight_norm=True
            ),
            slayer.block.cuba.Flatten(),
            slayer.block.cuba.Dense(neuron_dense_params, 8*8*32, 512, weight_norm=True),
            slayer.block.cuba.Dense(neuron_dense_params, 512, 11, weight_norm=True)
        ])
```
**Observed behavior:** 
- at the beginning of epoch 0 (so, during the very first iterations), my loss is large (like loss >= 75) but decreases quickly until it reaches something like `loss=17.6` (generally at the end of the first or second epoch). Then, it is completely stuck (so the grad flow completely shuts down to 0 at each layer).
- My first try was to remove put `weight_scale=1` in the conv layers and put a `weight_scale=2.5` in dense layers. The behavior of my dense layers was to become dead neurons (they did not spike anything) so nothing is learned.
- My second try was to inspire from your implementation on SlayerPytorch (https://github.com/bamsumit/slayerPytorch/blob/master/exampleLoihi/03_IBMGesture/ibmGestureTrain.ipynb) that use weight_scales of 10 and 50 for conv1 and conv2 respectively (it is the configuration showed in the code above). With this configuration, I do have some neural activity everywhere but the loss is stuck around 17.+ and there is no learning. My grad flow shows something like this:

![image](https://user-images.githubusercontent.com/32375997/164296274-5447322e-bfeb-4705-be9d-efa9d5305528.png)

I probably failed to reproduce your architecture here. Is there any value of parameters I can change ?
